File: 백준/Silver/2178. 미로 탐색/미로 탐색.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        int[][] miro = new int[N+1][M+1];
        boolean[][] visited = new boolean[N+1][M+1];

        for (int i = 1; i <= N; i++) {
            String line = br.readLine();
            char[] tmp_arr = line.toCharArray();
            for (int j = 1; j <= M; j++) {
                int input = Integer.parseInt(String.valueOf(tmp_arr[j-1]));
                miro[i][j] = input;
            }
        }

        int ans = bfs(miro, visited, N, M);

        bw.write(String.valueOf(ans));
        bw.newLine();
        bw.flush();
        bw.close();
    }

    private static int bfs(int[][] miro, boolean[][] visited, int N, int M) {
        Queue<int[]> myqueue = new LinkedList<>(); // y, x, move
        myqueue.add(new int[]{1,1,1});

        int[] dy = {-1, 0, 1, 0};
        int[] dx = {0, 1, 0, -1};

        while (!myqueue.isEmpty()) {
            int[] now = myqueue.poll();
            int now_y = now[0];
            int now_x = now[1];
            int now_move = now[2];

            if (now_y == N && now_x == M) {
                return now_move;
            }

            for (int k = 0; k < 4; k++) {
                int y = now_y + dy[k];
                int x = now_x + dx[k];

                if (1 <= y && y <= N && 1 <= x && x <= M &&
                        miro[y][x] == 1 && !visited[y][x]) {
                    visited[y][x] = true;
                    myqueue.add(new int[]{y,x,now_move+1});
                }
            }
        }

        return 0;
    }
}

File: 백준/Silver/2178. 미로 탐색/미로 탐색.cc
#include <iostream>
#include <vector>
#include <queue>
#include <tuple>
using namespace std;

int dy[4] = {-1, 0, 1, 0};
int dx[4] = {0, 1, 0, -1};
int N, M;
vector<vector<int>> map;

int bfs() {
    queue<tuple<int, int, int>> myqueue;
    vector<vector<bool>> visited(N+1, vector<bool>(M+1, false));

    // (1,1) 에서 출발
    myqueue.push(make_tuple(1,1,1));
    visited[1][1] = true;

    while (!myqueue.empty()) {
        tuple<int, int, int> now = myqueue.front();
        myqueue.pop();
        int nowY = get<0>(now);
        int nowX = get<1>(now);
        int path = get<2>(now);

        if (nowY == N && nowX == M) return path;

        for (int k = 0; k < 4; k++) {
            int nextY = nowY + dy[k];
            int nextX = nowX + dx[k];

            if (1 <= nextY && nextY <= N && 1 <= nextX && nextX <= M && !visited[nextY][nextX]) {
                if (map[nextY][nextX] == 1) {
                    visited[nextY][nextX] = true;
                    myqueue.push(make_tuple(nextY,nextX,path+1));
                }
            }
        }
    }

    return 0;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    cin >> N >> M;

    // 1. 맵 입력받기
    map.resize(N+1, vector<int>(M+1));
    for (int i = 1; i <= N; i++) {
        string now;
        cin >> now;
        for (int j = 1; j <= M; j++) {
            map[i][j] = now[j-1] - '0';
        }
    }

    // 2. 이동
    int ans = bfs();
    cout << ans;

    return 0;
}